design/authenticators/authn_gcp/authn_gcp_solution_design.md

```diff
@@ -0,0 +1,312 @@
+|     Mapping host annotation logic                                  |     -                                     |     Map each above key   to his location in the google JWT token                    |
+
+The following diagram represents JWT authenticator with a lot of small responsibility units, according to the above table and past experience, 
+in order to have the flexibility to support future vendors that uses JWT token for their authentication process, like google, GCP
+
+**Note:** in this feature we will support only GCP
+![Authn gcp class diagram](authn-gcp-class-diagram.png)
+
+GCP & JWT will have 3 dependencies which will change their behavior by dependency injection  
+
+1. **Fetch Authenticator configuration**
+
+    1.1 **Fetch provider uri**
+    
+    * GCP implementation - return an hardcoded
+    * Authenticator policy implementation - fetch from provider uri value in the policy, this logic is relevant for: OIDC, Azure, JWT
+    
+    1.2 **Fetch claims**
+        
+      1.2.1 **Fetch mandatory claims keys to validate** - list of claims that must exist in token
+        
+      * GCP implementation - return an hardcoded list (Iss, exp, aud, iat, nbf)
+      * OIDC needs refactor - return an hardcoded list (Iss, exp, nbf)
+      * Azure needs refactor - return an hardcoded list (Iss, exp, nbf)
+      * JWT design phase - return an list which defined in a variable under the authenticator policy
+      
+      1.2.2 **Fetch mandatory claims values** - values of claims to validate which can be reside in different location: hardcoded, policy, hostname, uri of the request
+      * GCP implementation - return an hardcoded values (Iss = "https://www.googleapis.com/oauth2/v1/certs", aud = "conjur"
+      * OIDC needs refactor - fetch iss value from the provider uri 
+      * Azure needs refactor - fetch iss value from the provider uri 
+      * JWT design phase - return the values of mandatory claims from variable in policy
+      
+      1.2.3 **Fetch optional claims to validate** - list of claims that we should validate only if they appears in the token 
+      * GCP - not relevant
+      * OIDC - not relevant
+      * Azure - not relevant
+      * JWT design phase - return the values of optional claims from variable in policy
+      
+      1.2.4 **Fetch optional claims values**
+      * JWT design phase - return the values of optional claims from variable in policy
+      
+      1.2.5 **authn prefix parameter** - the authenticator prefix which will later be use the fetch a the host annotations 
+
+2. **Validate and decode** - same code a today with small refactoring of sending also the claims values if necessary 
+    
+    * Fetch public key
+    * Validate signature
+    * Validate claims
+    * Save key in cache 
+
+3. **Validate resource restrictions** - Responsible to validate the resource restrictions which defined in conjur, the idea is to refactor Azure and K8S to work with this pattern 
+    
```

Comment:
It the Azure refactoring going to be a part of current project ?

Reply:
No, i want to suggest to do it at the moment we will finish this feature

Reply:
@nessiLahav - but you mean here that resource restriction validation for GCP will be a part of this feature, right ?

Reply:
Yes